lib/dialer/retrydialer.go

```diff
@@ -0,0 +1,135 @@
+	"time"
+)
+
+// NoCandidateUpstreams is an error returned by RetryDialer if there are no
+// upstreams given as candidates to dial.
+var NoCandidateUpstreams = errors.New("no candidate upstreams")
+var ConnectionTypeUnsupported = errors.New("connection type is not supported")
+
+// DialPolicy controls which upstream to dial, out of a set of candidates.
+//
+// Multiple goroutines may invoke methods on a UpstreamDialer simultaneously.
+type DialPolicy interface {
+
+	// ChooseBestUpstream asks the policy to choose an upstream from the
+	// given set of candidates. In the case that the policy decides
+	// none of the candidates are feasible, an error is returned.
+	ChooseBestUpstream(candidates core.UpstreamSet) (core.Upstream, error)
+
+	// DialFailed informs the policy that a dial attempt failed
+	DialFailed(upstream core.Upstream, symptom error)
+
+	// DialSucceeded informs the policy that a dial attempt succeeded
+	DialSucceeded(upstream core.Upstream)
+
+	// ConnectionClosed informs the policy that a connection created by a prior
+	// successful dial attempt has been closed.
+	ConnectionClosed(upstream core.Upstream)
+}
+
+// UpstreamDialer dials upstreams.
+//
+// Multiple goroutines may invoke methods on a UpstreamDialer simultaneously.
+type UpstreamDialer interface {
+	// DialUpstream dials upstream, returning a DuplexConn if a connection is established.
+	// Implementations should honour context deadlines, timeouts, and cancellations (if any).
+	DialUpstream(ctx context.Context, upstream core.Upstream) (forwarder.DuplexConn, error)
+}
+
+type SimpleUpstreamDialer struct{}
+
+func (d SimpleUpstreamDialer) DialUpstream(ctx context.Context, upstream core.Upstream) (forwarder.DuplexConn, error) {
+	dd := net.Dialer{}
+	conn, err := dd.DialContext(ctx, upstream.Network, upstream.Address)
+	if err != nil {
+		return nil, err
+	}
+	switch c := conn.(type) {
+	case *net.TCPConn:
+		return c, nil
+	case *tls.Conn:
+		return c, nil
+	default:
+		_ = conn.Close()
+		return nil, ConnectionTypeUnsupported
+	}
+}
+
+// RetryDialer attempts to dial a candidate Upstream as selected by a
+// configurable DialPolicy. If the dial attempt fails, it informs the policy
+// of the failure and asks the policy for the next candidate upstream.
+// RetryDialer requires a Timeout to be supplied, which is shared across
+// all dial attempts.
+//
+// Multiple goroutines may invoke methods on a RetryDialer simultaneously.
+type RetryDialer struct {
+	Logger      slog.Logger
+	Timeout     time.Duration // Timeout to apply for each DialBestUpstream operation.
+	Policy      DialPolicy
+	InnerDialer UpstreamDialer
+}
+
+func (d *RetryDialer) DialBestUpstream(ctx context.Context, candidates core.UpstreamSet) (core.Upstream, forwarder.DuplexConn, error) {
+	if len(candidates) == 0 {
+		return core.Upstream{}, nil, NoCandidateUpstreams
+	}
+	// TODO use shorter timeout for each of n > 1 dial attempts?
+	dialCtx, cancel := context.WithTimeout(ctx, d.Timeout)
```

Comment:
i think this is important in practice, otherwise if we spent the entire timeout on the first dial attempt, that's not very useful.

maybe a simple way to do it could be to cut the timeout in 2 or 3, and set that as a max timeout per inner dial. adds a little complexity as then there are two lots of timeouts to keep track of.

another idea is that we could dial multiple upstreams in parallel and see which one (if any) succeeds first.